### ISSUE TITLE
Bug in JTFS S1 shape and `subsample_fourier` during frequency scattering (numpy and tensorflow)

### ISSUE THREAD

**Issue by cyrusvahidi** [kymatio/kymatio#1029]
In time-frequency scattering we stack the S1 paths along the 2nd axis frequency:

https://github.com/kymatio/kymatio/blob/8f2bbdfdbe974de82caf52b07826859c02d3bd24/kymatio/scattering1d/core/timefrequency_scattering.py#L157 

This is correct in torch, since the default value for `axis=2`, however, `axis=1` in `numpy_backend` and `tensorflow_backend`, resulting in an array of shape `[batch, \lambda_1, 1, t]`. This has knock on effects for frequency scattering. This is a simple fix:

`S_1 = backend.stack(S_1_list, dim=2)`.

Another issue arises in `subsample_fourier` in tensorflow and numpy during frequency scattering:

https://github.com/kymatio/kymatio/blob/8f2bbdfdbe974de82caf52b07826859c02d3bd24/kymatio/scattering1d/core/timefrequency_scattering.py#L265 

Numpy and TF backends follow the same logic for reshaping:

https://github.com/kymatio/kymatio/blob/8f2bbdfdbe974de82caf52b07826859c02d3bd24/kymatio/scattering1d/backend/numpy_backend.py#L28 

https://github.com/kymatio/kymatio/blob/8f2bbdfdbe974de82caf52b07826859c02d3bd24/kymatio/scattering1d/backend/tensorflow_backend.py#L30

which ignores the extra first dimension. This has serious consequences during subsampling, as the time dimension is swallowed into the batch dimnension. 

The torch backend makes sure to handle all dimensions upto the last dimension that we're subsampling over:

https://github.com/kymatio/kymatio/blob/8f2bbdfdbe974de82caf52b07826859c02d3bd24/kymatio/scattering1d/backend/torch_backend.py#L46 

The numpy and tensorflow implementations can be easily fixed by fixing the reshaping in `subsample_fourier`.

This arose when implementing Keras frontend. I will make a PR that fixes the stacking of S1, the reshaping logic in np and tf `subsample_fourier` and implements the Keras frontend.

**Comment by eickenberg** [comment]
these should all be aligned and the backend should regularize everything to
numpy convention (which for stack means axis=0 by default)

**Comment by cyrusvahidi** [comment]
> these should all be aligned and the backend should regularize everything to numpy convention (which for stack means axis=0 by default)
> […](#)

Right. This turns out to not be the case. The default `dim` differs. Note that `stack` is defined in the top-level backend.

**Comment by eickenberg** [comment]
yeah I'm describing the desired state ;)

I'd like to go much further and expose basic numpy functions on all
backends with the same function signatures. this would unblock a lot of
development. but that is not for this release

**Comment by cyrusvahidi** [comment]
On further investigation, the reshaping that we do in `subsample_fourier` assumes that the input shape is `[batch, 1, time]` and swallows dim `1` into the batch dimension. Subsequent operations in `Scattering1D` assume shape of `[batch, time]`, which is returned by `subsample_fourier`. This extra dim is created in the top-level `backend.reshape_input` for 1D, 2D and 3D. 

This falls apart when we're doing frequency scattering, since we deal with shapes `[batch, time, freq]`. In `numpy_backend.subsample_fourier`, we do `y = x.reshape(-1, k, x.shape[-1] // k)`, hence the `time` dimension is lost in the batch dimension.

I propose that we fix the reshaping logic in 1D `subsample_fourier` and leave it to the scattering routine to handle the removal of the extra dimension, wherever necessary.

**Comment by lostanlen** [comment]
Agree with @cyrusvahidi . Note that 2D does not introduce this extra dimension

```python
        batch_shape = input.shape[:-2]
        signal_shape = input.shape[-2:]

        input = input.reshape((-1,) + signal_shape)

        S = scattering2d(input, self.pad, self.unpad, self.backend, self.J,
                self.L, self.phi, self.psi, self.max_order, self.out_type)
```

**Comment by MuawizChaudhary** [comment]
Hi, How has progress been on this issue? did we converge to a solution?
